apps/crawler/test/task_manager_test.exs

```
@@ -0,0 +1,101 @@
+defmodule TaskManagerTest do
+  use ExUnit.Case
+  require Logger
+  alias Crawler.TaskManager
+
+  defmodule EchoProvider do
+    def search(query) do
+      query
+    end
+  end
+
+  defmodule BlockingProvider do
+    def search(query) do
+      :timer.sleep(:infinity)
+      query
+    end
+  end
+
+  test "basic test ensure provider.search(query) gets called" do
+    input = ["basic testvalue"]
+    {:ok, mng} = TaskManager.start_link(1)
+    result = TaskManager.search(input, mng, EchoProvider)
+    assert result == input
+    Process.sleep(400)
+  end
+
+  test "don't hang on unending tasks" do
+    {:ok, mng} = TaskManager.start_link(3)
+    input = ["unending testvalue"]
+
+    timeout = 1000
+
+    block = fn ->
+      {pid, ref} =
+        Process.spawn(
+          fn ->
+            TaskManager.search(:lemming, mng, BlockingProvider, 500)
+          end,
+          [:monitor]
+        )
+
+      receive do
+        {:DOWN, ^ref, :process, ^pid, {reason, _more_info}} -> reason
+      end
+    end
+
+    valid = fn ->
+      try do
+        TaskManager.search(input, mng, EchoProvider, timeout)
+      catch
+        :exit, msg = {reason, _} ->
+          Logger.warn("valid request failed with #{inspect(msg)}")
```

it's probably better to raise here rather than simply log, since this should never happen

---

well, it should happen if TaskManager.search exits ungracefully.
in that case reason is never `:ok` and the assert fails.
this code is just to keep the failing assertion clearly and the full exit details in the log.